### ISSUE TITLE
System.InvalidCastException when casting db2 boolean

### ISSUE THREAD

**Issue by JonasHansen16** [DapperLib/Dapper#1880]
We recently upgraded our db2client for .net from v11.1.3.3 to v11.5.8.0
With the upgrade we are getting problems with mapping db2 bool columns.

We are now getting the following error

> System.Data.DataException: Error parsing column 3 (EXAMPLEBOOL=1 - Int16) ---> System.InvalidCastException: Specified cast is not valid.
> 
>    at Deserialized25043f1-669b-4d9f-be69-ae57be5bf3c7(IDataReader )
>    --- End of inner exception stack trace ---
>    at Dapper.SqlMapper.ThrowDataException(Exception ex, Int32 index, IDataReader reader, Object value) in /_/Dapper/SqlMapper.cs:line 3706
>    at Deserialized25043f1-669b-4d9f-be69-ae57be5bf3c7(IDataReader )
>    at Dapper.SqlMapper.<QueryImpl>d__140`1.MoveNext() in /_/Dapper/SqlMapper.cs:line 1113
>    at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
>    at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
>    at Dapper.SqlMapper.Query[T](IDbConnection cnn, String sql, Object param, IDbTransaction transaction, Boolean buffered, Nullable`1 commandTimeout, Nullable`1 commandType) in /_/Dapper/SqlMapper.cs:line 734
>    at BoolMappingTest.Example.GetExampleEntities() in E:\Builds\BA02\_work\1985\s\BoolMappingTest\Example.cs:line 33
>    at per.personeelsdata.api.Controllers.PersoneelController.GetExampleEntities() in E:\Builds\BA02\_work\1985\s\per.personeelsdata.api\Controllers\PersoneelController.cs:line 143
>    at lambda_method(Closure , Object , Object[] )
>    at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass10.<GetExecutor>b__9(Object instance, Object[] methodParameters)
>    at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken cancellationToken)
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__0.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__5.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__5.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__0.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__2.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__2.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
>    at UH.ISA.Api.Core.Authentication.AddChallengeOnUnauthorizedResult.<ExecuteAsync>d__9.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Controllers.AuthenticationFilterResult.<ExecuteAsync>d__0.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at System.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__1.MoveNext()

We've brought this up with IBM, they gave us the following explanation.

> Found that DB2 .NET provider of db2 client v11.1.3.3, is getting Int16 data type for boolean column (i.e examplebool) in sql query from underneath native layer as shown below. And dapper is able to process this Int16 type without any issue.
However, DB2 .NET provider of db2 client version 11.5.8.0, is receiving boolean data type for boolean column from underneath layer. And dapper is unable to process boolean type and throws System.InvalidCastException exception.

Currently we are sidestepping the problem by doing an explicit cast in the query itself.
Would love to know your thoughts on this.

Thank you in advance

**Comment by mgravell** [comment]
I might be wrong, but it sounds like they've inverted things there. We can see from the error message that Dapper is being presented an Int16 value. If your model has bool, then indeed Int16 doesn't cast cleanly to bool. Am I misunderstanding? Could you perhaps give more example of the query and model being used here? And in particular the cast in the query that you say works around it? Note I'm not a DB2 expert - I can only comment on what ADO.NET gives us and what we do with it

**Comment by JonasHansen16** [comment]
@mgravell 
I wrote up a short example for IBM to reproduce the issue.

the sql is as follows
`var sql = "select 'exampleText' as exampleText, 2022 as exampleint, 0 as exampleSmallInt, true as examplebool from sysibm.sysdummy1 ";`

the model to map to.

```
public class ExampleEntity
    {
        public string TextField { get; set; }
        public bool SmallIntField { get; set; }
        public int IntegerField { get; set; }
        public bool BoolField { get; set; }
    }
```
The property SmallIntField maps correctly to the boolean.
In previous versions of the db2client the property BoolField also mapped correctly.

But now we get the error.

The workaround in the query would be to cast the examplebool as such
`integer(true) as examplebool`

**Comment by mgravell** [comment]
Well, I wouldn't expect that map and query to do *anything* interesting, because none of the column names match the properties, so: Dapper isn't going to touch any of them. But: I'll assume that's just copy/paste

Is there any chance you can run the following and post the output?

``` c#
using var cmd = connection.CreateCommand();
cmd.Connection = connection; // probably redundant, but: make sure
cmd.CommandText = "select 'exampleText' as exampleText, 2022 as exampleint, 0 as exampleSmallInt, true as examplebool from sysibm.sysdummy1 ";
cmd.CommandType = CommandType.Text;
using var reader = cmd.ExecuteReader();
// show the declared types before we see the row (if this fails, just remove this)
for (int i = 0; i < reader.FieldCount; i++)
{
    Console.WriteLine($"{reader.GetName(i)}: {reader.GetFieldType(i).FullName}");
}

if (!reader.Read()) throw new InvalidOperationException("expected a row!");
// show the actual types
for (int i = 0; i < reader.FieldCount; i++)
{
    Console.WriteLine($"{reader.GetName(i)}: {reader.GetValue(i)?.GetType().FullName}");
}
// show the declared types after we've seen the row
for (int i = 0; i < reader.FieldCount; i++)
{
    Console.WriteLine($"{reader.GetName(i)}: {reader.GetFieldType(i).FullName}");
}
```

**Comment by JonasHansen16** [comment]
I'm using [Fluentmap ](https://github.com/henkmollema/Dapper-FluentMap) to handle the mapping of the columns to the right properties.
Not very useful in this example but handy in our actual code.

to map the properties for the example i used the following class.
public class MappingExample : EntityMap<ExampleEntity>

```
        public MappingExample()
        {
            Map(p => p.TextField)
                .ToColumn("EXAMPLETEXT");
            Map(p => p.SmallIntField)
                .ToColumn("EXAMPLEINT");
            Map(p => p.IntegerField)
                .ToColumn("EXAMPLESMALLINT");
            Map(p => p.BoolField)
                .ToColumn("EXAMPLEBOOL");
        }
    }
```

On the old version the output is:

> EXAMPLETEXT: System.String
> EXAMPLEINT: System.Int32
> EXAMPLESMALLINT: System.Int32
> EXAMPLEBOOL: System.Int16
> EXAMPLETEXT: System.String
> EXAMPLEINT: System.Int32
> EXAMPLESMALLINT: System.Int32
> EXAMPLEBOOL: System.Int16
> EXAMPLETEXT: System.String
> EXAMPLEINT: System.Int32
> EXAMPLESMALLINT: System.Int32
> EXAMPLEBOOL: System.Int16

On the new version the output is:

> EXAMPLETEXT: System.String
EXAMPLEINT: System.Int32
EXAMPLESMALLINT: System.Int32
EXAMPLEBOOL: System.Boolean
EXAMPLETEXT: System.String
EXAMPLEINT: System.Int32
EXAMPLESMALLINT: System.Int32
EXAMPLEBOOL: System.Int16
EXAMPLETEXT: System.String
EXAMPLEINT: System.Int32
EXAMPLESMALLINT: System.Int32
EXAMPLEBOOL: System.Boolean

**Comment by mgravell** [comment]
OK, so this is a bug in db2, and should be reported to them; they are *declaring* `EXAMPLEBOOL` as `bool`, but that isn't what they're giving us; they're *giving* us a `short`:

- `EXAMPLEBOOL: System.Boolean` <== initial declaration
- `EXAMPLEBOOL: System.Int16` <== actual value
- `EXAMPLEBOOL: System.Boolean` <== declaration after seeing a row

If they *tell us* that they're giving us a `bool`, but they *actually* give us a `short`: we're going to perform an invalid cast. Please report this to db2. This is **their bug**. If you show them the above printout, they should understand the problem.

**Comment by JonasHansen16** [comment]
Thank you for helping clarify this for me.
I'll take this back up with IBM.

**Comment by JonasHansen16** [comment]
For posterity's sake.
The issue was indeed with the db2 client.
They told us this issue will be resolved in the next major release (v12.1).

**Comment by mgravell** [comment]
Thanks for the confirmation/ update
